### ISSUE TITLE
dllist.h:201:11: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer

### ISSUE THREAD

**Issue by indygreg** [microsoft/snmalloc#292]
Thank you for snmalloc. I'm using it in https://github.com/indygreg/PyOxidizer (thanks to @ryancinsight's work) as an alternative memory allocator for embedded Python interpreters and it seems to have a meaningful impact on performance for some workloads.

If you attempt to build snmalloc targeting an older macOS deployment target, you get a compiler error due to features requiring macOS 10.14. The following build log is via Rust, but the failure is in the C library, so reporting it here.

```
  running: "cmake" "--build" "." "--target" "snmallocshim-1mib-rust" "--config" "Release" "--"
  Scanning dependencies of target snmallocshim-1mib-rust
  [100%] Building CXX object CMakeFiles/snmallocshim-1mib-rust.dir/src/override/rust.cc.o

  --- stderr
  CMake Warning at CMakeLists.txt:79 (message):
    Not generating clangformat target, no clang-format tool found
  Call Stack (most recent call first):
    CMakeLists.txt:361 (clangformat_targets)


  CMake Warning:
    Manually-specified variables were not used by the project:

      CMAKE_ASM_COMPILER
      CMAKE_ASM_FLAGS


  In file included from /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/rust.cc:2:
  In file included from /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/malloc.cc:1:
  In file included from /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/slowalloc.h:3:
  In file included from /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/globalalloc.h:4:
  In file included from /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/alloc.h:12:
  In file included from /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/chunkmap.h:7:
  In file included from /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/mediumslab.h:3:
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/../ds/dllist.h:201:11: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
            delete c;
            ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/../ds/dllist.h:73:7: note: in instantiation of member function 'snmalloc::DLList<snmalloc::Mediumslab, nullptr_t, false>::clear' requested here
        clear();
        ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/alloc.h:664:5: note: in instantiation of member function 'snmalloc::DLList<snmalloc::Mediumslab, nullptr_t, false>::~DLList' requested here
      Allocator(
      ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/largealloc.h:250:22: note: in instantiation of member function 'snmalloc::Allocator<&snmalloc::needs_initialisation, &snmalloc::init_thread_allocator, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<20, unsigned char> > >, true>::Allocator' requested here
        return new (p) T(std::forward<Args...>(args)...);
                       ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/pool.h:61:23: note: in instantiation of function template specialization 'snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >::alloc_chunk<snmalloc::Allocator<&snmalloc::needs_initialisation, &snmalloc::init_thread_allocator, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<20, unsigned char> > >, true>, 4096, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> > &>' requested here
              .template alloc_chunk<T, bits::next_pow2_const(sizeof(T))>(
                        ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/globalalloc.h:34:22: note: in instantiation of function template specialization 'snmalloc::Pool<snmalloc::Allocator<&snmalloc::needs_initialisation, &snmalloc::init_thread_allocator, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<20, unsigned char> > >, true>, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> > >::acquire<snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> > &>' requested here
        return Parent::acquire(Parent::memory_provider);
                       ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/slowalloc.h:25:51: note: in instantiation of member function 'snmalloc::AllocPool<snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >, snmalloc::Allocator<&snmalloc::needs_initialisation, &snmalloc::init_thread_allocator, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<20, unsigned char> > >, true> >::acquire' requested here
      SlowAllocator() : alloc(current_alloc_pool()->acquire()) {}
                                                    ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/../ds/dllist.h:201:11: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
            delete c;
            ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/../ds/dllist.h:201:11: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
            delete c;
            ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/../ds/dllist.h:73:7: note: in instantiation of member function 'snmalloc::DLList<snmalloc::Superslab, nullptr_t, false>::clear' requested here
        clear();
        ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/alloc.h:664:5: note: in instantiation of member function 'snmalloc::DLList<snmalloc::Superslab, nullptr_t, false>::~DLList' requested here
      Allocator(
      ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/largealloc.h:250:22: note: in instantiation of member function 'snmalloc::Allocator<&snmalloc::needs_initialisation, &snmalloc::init_thread_allocator, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<20, unsigned char> > >, true>::Allocator' requested here
        return new (p) T(std::forward<Args...>(args)...);
                       ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/pool.h:61:23: note: in instantiation of function template specialization 'snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >::alloc_chunk<snmalloc::Allocator<&snmalloc::needs_initialisation, &snmalloc::init_thread_allocator, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<20, unsigned char> > >, true>, 4096, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> > &>' requested here
              .template alloc_chunk<T, bits::next_pow2_const(sizeof(T))>(
                        ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/globalalloc.h:34:22: note: in instantiation of function template specialization 'snmalloc::Pool<snmalloc::Allocator<&snmalloc::needs_initialisation, &snmalloc::init_thread_allocator, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<20, unsigned char> > >, true>, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> > >::acquire<snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> > &>' requested here
        return Parent::acquire(Parent::memory_provider);
                       ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/slowalloc.h:25:51: note: in instantiation of member function 'snmalloc::AllocPool<snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >, snmalloc::Allocator<&snmalloc::needs_initialisation, &snmalloc::init_thread_allocator, snmalloc::MemoryProviderStateMixin<snmalloc::PALApple<241> >, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<20, unsigned char> > >, true> >::acquire' requested here
      SlowAllocator() : alloc(current_alloc_pool()->acquire()) {}
                                                    ^
  /Users/gps/.cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.2.25/snmalloc/src/override/../mem/../ds/dllist.h:201:11: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
            delete c;
            ^
  2 errors generated.
  make[3]: *** [CMakeFiles/snmallocshim-1mib-rust.dir/src/override/rust.cc.o] Error 1
  make[2]: *** [CMakeFiles/snmallocshim-1mib-rust.dir/all] Error 2
  make[1]: *** [CMakeFiles/snmallocshim-1mib-rust.dir/rule] Error 2
```

I'm sympathetic to not supporting old macOS versions to make your project maintenance simpler. However, even the modern macOS 11.x SDKs still support targeting down to macOS 10.9 and it is still a common practice for applications to target old, hardly used macOS versions because it is often possible with minimal extra effort and it avoids angering users clinging to these old releases. PyOxidizer is designed to target down to 10.9. So I would gently encourage rethinking the minimum macOS support requirement.

That being said, if snmalloc wishes to impose a minimum macOS targeting requirement, I won't protest that decision: I just need to know what constraints to impose in PyOxidizer when snmalloc is being used. (As a former maintainer of Firefox's build system, I know the pain of supporting old operating systems and support whatever decision you make.)

**Comment by mjp41** [comment]
@indygreg, that code is unreachable in `snmalloc`, and was added to enable reuse for the list class in another project.  I think we should be able to fix it.  The general rule we have is to support platforms that are still getting security updates.

@ashamis, is we refactor `dllist.h` to take a template parameter of what to do with the object on clear, rather than calling `delete`, would that still work for your use case?  I would like not to call `delete` inside the allocator code, as that would be bad.

**Comment by mjp41** [comment]
So I am guessing adding a `constexpr` to the following `if`

https://github.com/microsoft/snmalloc/blob/35346e72c36e5208b2bb0cb2631a3e3ed7e68fc2/src/ds/dllist.h#L199-L202

would fix the issue, as that code is not considered compiled.

**Comment by ashamis** [comment]
That should not be a problem for me

**Comment by ashamis** [comment]
> @indygreg, that code is unreachable in `snmalloc`, and was added to enable reuse for the list class in another project. I think we should be able to fix it. The general rule we have is to support platforms that are still getting security updates.
> 
> @ashamis, is we refactor `dllist.h` to take a template parameter of what to do with the object on clear, rather than calling `delete`, would that still work for your use case? I would like not to call `delete` inside the allocator code, as that would be bad.

That should not be a problem for me

**Comment by mjp41** [comment]
@indygreg, we are using C++17 features fairly heavily in the codebase, so to run on 10.9, it will need to use a fairly recent version of clang/gcc.  

In #293, I have hopefully fixed the issue you found, but I don't have a CI process to check if anything else will fail.  Can you test fixes before they appear in snmalloc-rs?  I would sooner not cause the snmalloc-rs project (@SchrodingerZhu) too much work if possible.

**Comment by mjp41** [comment]
@indygreg, is it okay to close this issue.  I believe the specific problem you reported is addressed, but have new related issues appeared?

**Comment by mjp41** [comment]
Closed by #293
